src/ethereum/frontier/vm/instructions/environment.py

```diff
@@ -0,0 +1,244 @@
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_BASE)
+    push(evm.stack, U256.from_be_bytes(evm.current))
+
+
+def balance(evm: Evm) -> None:
+    """
+    Pushes the balance of the given account onto the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `1`.
+    OutOfGasError
+        If `evm.gas_left` is less than `20`.
+    """
+    # TODO: There are no test cases against this function. Need to write
+    # custom test cases.
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_EXT)
+
+    address = pop(evm.stack).to_be_bytes()
+    if len(address) > MAX_ADDRESS_BYTES:
+        # Get the last 20 bytes and consider this as the address
+        address = address[-MAX_ADDRESS_BYTES:]
+
+    account_state = evm.env.state.get(address, None)
+    if account_state is None:
+        balance = U256(0)
+    else:
+        balance = U256(account_state.balance)
+
+    push(evm.stack, balance)
+
+
+def origin(evm: Evm) -> None:
+    """
+    Pushes the address of the original transaction sender to the stack.
+    The origin address can only be an EOA.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    OutOfGasError
+        If `evm.gas_left` is less than `2`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_BASE)
+    push(evm.stack, U256.from_be_bytes(evm.env.origin))
+
+
+def caller(evm: Evm) -> None:
+    """
+    Pushes the address of the caller or transaction sender. The caller can be
+    an EOA or another CA.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    OutOfGasError
+        If `evm.gas_left` is less than `2`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_BASE)
+    push(evm.stack, U256.from_be_bytes(evm.caller))
+
+
+def callvalue(evm: Evm) -> None:
+    """
+    Obtain the value included in the transaction responsible for this
+    execution. Also pushes this value onto the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    OutOfGasError
+        If `evm.gas_left` is less than `2`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_BASE)
+    push(evm.stack, evm.value)
+
+
+def calldataload(evm: Evm) -> None:
+    """
+    Push the input data of current environment onto the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `1`.
+    OutOfGasError
+        If `evm.gas_left` is less than `3`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
+
+    # Converting start_index to Uint from U256 as start_index + 32 can
+    # overflow U256.
+    start_index = Uint(pop(evm.stack))
+    calldataload = evm.data[start_index : start_index + 32]
```

Comment:
I would prefer if we avoid shadowing `calldataload` with a different type, though I might be being a tiiiiiny bit pedantic.

Reply:
Done.